system/lib/wasmfs/file_table.cpp

```diff
@@ -0,0 +1,115 @@
+// Copyright 2021 The Emscripten Authors.  All rights reserved.
+// Emscripten is available under two separate licenses, the MIT license and the
+// University of Illinois/NCSA Open Source License.  Both these licenses can be
+// found in the LICENSE file.
+// This file defines the open file table of the new file system.
+// Current Status: Work in Progress.
+// See https://github.com/emscripten-core/emscripten/issues/15041.
+
+#include "file_table.h"
+
+namespace wasmfs {
+
+std::vector<std::shared_ptr<OpenFileState>> FileTable::entries;
+
+static __wasi_errno_t writeStdBuffer(const uint8_t* buf, __wasi_size_t len,
+  void (*console_write)(const char*), std::vector<char>& fd_write_buffer) {
+  for (__wasi_size_t j = 0; j < len; j++) {
+    uint8_t current = buf[j];
+    if (current == '\0' || current == '\n') {
+      fd_write_buffer.push_back('\0'); // for null-terminated C strings
+      console_write(&fd_write_buffer[0]);
+      fd_write_buffer.clear();
+    } else {
+      fd_write_buffer.push_back(current);
+    }
+  }
+  return __WASI_ERRNO_SUCCESS;
+}
+
+class StdinFile : public File {
+
+  __wasi_errno_t write(const uint8_t* buf, __wasi_size_t len) override {
+    return __WASI_ERRNO_INVAL;
+  }
+
+  __wasi_errno_t read(const uint8_t* buf, __wasi_size_t len) override {
+    return __WASI_ERRNO_INVAL;
+  };
+};
+
+class StdoutFile : public File {
+  static std::vector<char> writeBuffer;
+
+  __wasi_errno_t write(const uint8_t* buf, __wasi_size_t len) override {
+    return writeStdBuffer(buf, len, &emscripten_console_log, writeBuffer);
+  }
+
+  __wasi_errno_t read(const uint8_t* buf, __wasi_size_t len) override {
+    return __WASI_ERRNO_INVAL;
+  };
+};
+
+class StderrFile : public File {
+  static std::vector<char> writeBuffer;
```

Comment:
Why are these vectors `static` ?

Reply:
Can't `StderrFile` and `StdoutFile` both share everything except the actual write methods here?  Can they both inherit from some intermediate bufferedOutputFile?  Then the `writeStdBuffer` function above could be method on that class?

Reply:
This re-factoring can also be left to the next PR since I am planning on moving `write` and `read` out of File.